kamu-core/src/infra/repos/dataset_repository_s3.rs

```diff
@@ -0,0 +1,473 @@
+        let dataset_key_prefix = self.s3_context.get_key(dataset_name.as_str());
+        let list_response = self
+            .s3_context
+            .client
+            .list_objects(ListObjectsRequest {
+                bucket: self.s3_context.bucket.clone(),
+                prefix: Some(dataset_key_prefix),
+                ..ListObjectsRequest::default()
+            })
+            .await
+            .int_err()?;
+
+        if let Some(contents) = list_response.contents {
+            self.s3_context
+                .client
+                .delete_objects(DeleteObjectsRequest {
+                    bucket: self.s3_context.bucket.clone(),
+                    delete: Delete {
+                        objects: contents
+                            .iter()
+                            .map(|obj| ObjectIdentifier {
+                                key: obj.key.clone().unwrap(),
+                                version_id: None,
+                            })
+                            .collect(),
+                        quiet: Some(true),
+                    },
+                    ..DeleteObjectsRequest::default()
+                })
+                .await
+                .int_err()?;
+        }
+
+        Ok(())
+    }
+
+    async fn finish_create_dataset(
+        &self,
+        dataset: &dyn Dataset,
+        staging_name: &str,
+        dataset_name: &DatasetName,
+    ) -> Result<DatasetHandle, CreateDatasetError> {
+        let summary = match dataset.get_summary(GetSummaryOpts::default()).await {
+            Ok(s) => Ok(s),
+            Err(GetSummaryError::EmptyDataset) => unreachable!(),
+            Err(GetSummaryError::Access(e)) => Err(e.int_err().into()),
+            Err(GetSummaryError::Internal(e)) => Err(CreateDatasetError::Internal(e)),
+        }?;
+
+        let handle = DatasetHandle::new(summary.id, dataset_name.clone());
+
+        match self
+            .move_bucket_items_on_dataset_rename(staging_name, dataset_name.as_str())
+            .await
+        {
+            Ok(_) => Ok(handle),
+            Err(MoveBucketItemsOnRenameError::NameCollision(e)) => {
+                Err(CreateDatasetError::NameCollision(e))
+            }
+            Err(MoveBucketItemsOnRenameError::Internal(e)) => Err(CreateDatasetError::Internal(e)),
+        }
+    }
+
+    async fn bucket_path_exists(&self, key_prefix: &str) -> Result<bool, InternalError> {
+        let listing = self
+            .s3_context
+            .client
+            .list_objects(ListObjectsRequest {
+                bucket: self.s3_context.bucket.clone(),
+                prefix: Some(self.s3_context.get_key(key_prefix)),
+                max_keys: Some(1),
+                ..ListObjectsRequest::default()
+            })
+            .await;
+
+        match listing {
+            Ok(resp) => Ok(resp.contents.is_some()),
+            Err(e) => Err(e.int_err().into()),
+        }
+    }
+
+    async fn move_bucket_items_on_dataset_rename(
```

Comment:
I feel like there's too much inconsistency between functions taking dataset names as parameters and those that take S3 paths / prefixes.

I would recommend creating a utility module for S3 with list / recursive delete / recursive move functionality and then using it here. Such module can probably unify some code with `ObjectStoreS3` too.